### ISSUE TITLE
Flask Oauth2 Auth Server: Error trying to decode a non urlencoded string. But it was encoded in the request

### ISSUE THREAD

**Issue by Avantol13** [lepture/authlib#93]
`authlib.flask.oauth2.authorization_server._create_oauth2_request` inits `OAuth2Request()` with `q.url`, which ends up *not* keeping percent-encoding for spaces (so instead of the `%20` from the url, it becomes an actual space ` `). `q` being a Flask request in the context of that function.

Issue shows up when passing `scope=openid%20user` or the like.
`GET {{ROOT}}/oauth2/authorize?response_type=code&client_id={{CLIENT_ID}}&redirect_uri={{REDIRECT}}&scope=openid%20user`

Trace:
```
Traceback (most recent call last):
  File "virtual-env/local/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "virtual-env/local/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "{{OUR_APP}}", line 83, in authorize
    grant = server.validate_consent_request(end_user=user)
  File "virtual-env/local/lib/python2.7/site-packages/authlib/flask/oauth2/authorization_server.py", line 208, in validate_consent_request
    req = _create_oauth2_request(request)
  File "virtual-env/local/lib/python2.7/site-packages/authlib/flask/oauth2/authorization_server.py", line 274, in _create_oauth2_request
    q.headers
  File "virtual-env/local/lib/python2.7/site-packages/authlib/specs/rfc6749/wrappers.py", line 39, in __init__
    self.query_params = url_decode(self.query)
  File "virtual-env/local/lib/python2.7/site-packages/authlib/common/urls.py", line 64, in url_decode
    raise ValueError(error % (set(query) - urlencoded, query))
ValueError: Error trying to decode a non urlencoded string. Found invalid characters: set([u' ']) in the string: 'response_type=code&client_id={{CLIENT_ID}}&redirect_uri={{REDIRECT}}&scope=openid user'. Please ensure the request/response body is x-www-form-urlencoded.
```

Currently Using:
authlib==0.9
flask==0.10.1
Werkzeug==0.12.2

NOTE: Our server class...
```
class OIDCServer(AuthorizationServer):
    """
    Implement the OIDC provider to attach to the flask app.

    Specific OAuth grants (authorization code, refresh token) are added on to
    a server instance using ``OIDCServer.register_grant_endpoint(grant)``. For
    usage, see ``fence/oidc/server.py``.
    """

    def init_app(self, app, query_client=None, save_token=None):
        if query_client is not None:
            self.query_client = query_client
        if save_token is not None:
            self.save_token = save_token
        self.app = app
        self.generate_token = generate_token
        self.init_jwt_config(app)
        if getattr(self, "query_client"):
            self.authenticate_client = ClientAuthentication(query_client)
```

It appears that in `werkzeug`, there's a `werkzeug.urls.url_to_iri` function that gets called to set the `url` attribute in the request (here: `werkzeug.wrappers.BaseRequest.url`) and at this point, the `%20` becomes a space ` `.

See: http://werkzeug.pocoo.org/docs/0.14/wrappers/#werkzeug.wrappers.BaseRequest.url

That `url` then gets passed into `OAuth2Request` in `authlib.flask.oauth2.authorization_server._create_oauth2_request` which eventually causes the issue noted above.

I assume that in `_create_oauth2_request` we should reconstruct the url without using the `.url` and instead combining perhaps `q.url_root` + `q.path` + `q.query_string` ? Note that `q.query_string` *does* maintain the `%20` without replacing it with a space.

Hopefully this has made sense, let me know if you want/need additional information.

**Comment by Avantol13** [comment]
note that replacing `q.url` with `q.url.replace(" ", "%20")` (or even `q.url.replace(" ", "+")`) fixes the issue. However, since this is an internal function, our class cannot easily override it

**Comment by Avantol13** [comment]
NOTE: I was able to work around the issue by modifying the flask request and passing it into the functions. Our new class:
```
class OIDCServer(AuthorizationServer):
    """
    Implement the OIDC provider to attach to the flask app.

    Specific OAuth grants (authorization code, refresh token) are added on to
    a server instance using ``OIDCServer.register_grant_endpoint(grant)``. For
    usage, see ``fence/oidc/server.py``.
    """

    def init_app(self, app, query_client=None, save_token=None):
        if query_client is not None:
            self.query_client = query_client
        if save_token is not None:
            self.save_token = save_token
        self.app = app
        self.generate_token = generate_token
        self.init_jwt_config(app)
        if getattr(self, "query_client"):
            self.authenticate_client = ClientAuthentication(query_client)

    def validate_consent_request(self, *args, **kwargs):
        request = kwargs.pop("request", None)
        kwargs["request"] = _ensure_spaces_percent_encoded(request)
        return super(OIDCServer, self).validate_consent_request(*args, **kwargs)

    def create_token_response(self, *args, **kwargs):
        request = kwargs.pop("request", None)
        kwargs["request"] = _ensure_spaces_percent_encoded(request)
        return super(OIDCServer, self).create_token_response(*args, **kwargs)

    def create_endpoint_response(self, *args, **kwargs):
        request = kwargs.pop("request", None)
        kwargs["request"] = _ensure_spaces_percent_encoded(request)
        return super(OIDCServer, self).create_endpoint_response(*args, **kwargs)


def _ensure_spaces_percent_encoded(request):
    q = request or flask_req
    q.url = q.url.replace(" ", "%20")
    return q
```

**Comment by valintepes** [comment]
I had a stack overflow post that was related to this that was answered and fixed. Or worked-around anyways.  The fix is, I think but am not sure, not yet in any of the releases.  It'll probably in 0.10.  Here's the [link to his answer](https://stackoverflow.com/questions/52300448/authlib-reports-error-trying-to-decode-a-non-urlencoded-string-when-using-andr/52344556#52344556)
and just for laziness's sake, [link to the commit](https://github.com/lepture/authlib/commit/d8ab09fb97169fc47070f48c2ede43348f1feff0)

**Comment by lepture** [comment]
This is a known issue. Fixed in 0.10.
